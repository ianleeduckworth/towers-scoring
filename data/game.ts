export type Player = {
    got?: number;
    bet?: number;
}

export type Round = {
    roundNumber: number;
    player1: Player;
    player2: Player;
    player3?: Player;
    player4?: Player;
}

export type Game = {
    players: {
        player1: string;
        player2: string;
        player3?: string;
        player4?: string;
    },
    rounds: Round[];
}