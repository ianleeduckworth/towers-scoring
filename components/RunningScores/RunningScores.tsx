import * as React from 'react';
import { StyleSheet, View, Text } from 'react-native';
import { Game } from '../../data/game';
import { defaults } from '../../styles/global';
import { calculateScore } from '../../utilities/scoreCalculator';

interface RunningScoreProps {
    game: Game;
}

export const RunningScores = ({ game }: RunningScoreProps) => {
    let player1Score = game.rounds.map(r =>
        (r.player1.bet !== undefined && r.player1.got !== undefined) ? calculateScore(r.player1.bet, r.player1.got) : 0
    ).reduce((a, b) => a + b);

    let player2Score = game.rounds.map(r =>
        (r.player2.bet !== undefined && r.player2.got !== undefined) ? calculateScore(r.player2.bet, r.player2.got) : 0
    ).reduce((a, b) => a + b);

    let player3Score = game.players.player3 ? game.rounds.map(r =>
        (r.player3?.bet !== undefined && r.player3.got !== undefined) ? calculateScore(r.player3.bet as number, r.player3.got as number) : 0
    ).reduce((a, b) => a + b) : undefined;

    let player4Score = game.players.player4 ? game.rounds.map(r =>
        (r.player4?.bet !== undefined && r.player4.got !== undefined) ? calculateScore(r.player4.bet as number, r.player4.got as number) : 0
    ).reduce((a, b) => a + b) : undefined;

    return (
        <View style={styles.container}>
            <View style={styles.titleContainer}>
                <Text style={styles.title}>Scores</Text>
            </View>
            <View style={styles.scoreContainer}>
                <View style={styles.scoreItem}>
                    <Text style={styles.text}>{game.players.player1}</Text>
                    <Text style={styles.text}>{player1Score}</Text>
                </View>
                <View style={styles.scoreItem}>
                    <Text style={styles.text}>{game.players.player2}</Text>
                    <Text style={styles.text}>{player2Score}</Text>
                </View>
                {player3Score &&
                    <View style={styles.scoreItem}>
                        <Text style={styles.text}>{game.players.player3}</Text>
                        <Text style={styles.text}>{player3Score}</Text>
                    </View>
                }
                {player4Score &&
                    <View style={styles.scoreItem}>
                        <Text style={styles.text}>{game.players.player4}</Text>
                        <Text style={styles.text}>{player4Score}</Text>
                    </View>
                }
            </View>
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        width: '100%',
        height: 100,
        backgroundColor: defaults.mediumGray,
        paddingLeft: 20,
        paddingRight: 20,
        borderWidth: 1,
        borderTopColor: defaults.black,
        borderBottomColor: defaults.black,
    },
    title: {
        color: defaults.black,
        fontWeight: 'bold',
        fontSize: defaults.titleTextSize,
        marginBottom: 4,
    },
    text: {
        fontSize: defaults.normalTextSize,
        fontFamily: defaults.font,
        color: defaults.black,
        fontWeight: 'bold',
    },
    titleContainer: {
        marginBottom: 3,
    },
    scoreContainer: {
        flexDirection: 'row',
    },
    scoreItem: {
        flexDirection: 'column',
        paddingRight: 40
    }
});