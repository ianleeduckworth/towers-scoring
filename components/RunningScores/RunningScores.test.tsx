import renderer from 'react-test-renderer';
import * as React from 'react';
import { RunningScores } from './RunningScores';
import { Game } from '../../data/game';

const baseGame: Game = {
  players: {
    player1: 'Player1',
    player2: 'Player2',
  },
  rounds: [{
    roundNumber: 1,
    player1: {
      bet: 20,
      got: 30,
    },
    player2: {
      bet: 20,
      got: 30,
    },
  }],
};

describe('button', () => {
  it('renders correctly with 2 players and 1 round', () => {
    const element = renderer.create(
            <RunningScores game={baseGame} />,
    ).toJSON();

    expect(element).toMatchSnapshot();
  });

  it('renders correctly with 3 players and 1 round', () => {
    const game = { ...baseGame };
    game.players.player3 = 'Player3';
    game.rounds[0].player3 = {
      bet: 20,
      got: 20,
    };

    const element = renderer.create(
            <RunningScores game={game} />,
    ).toJSON();

    expect(element).toMatchSnapshot();
  });

  it('renders correctly with 4 players and 1 round', () => {
    const game = { ...baseGame };
    game.players.player3 = 'Player3';
    game.players.player4 = 'Player4';
    game.rounds[0].player3 = {
      bet: 20,
      got: 20,
    };
    game.rounds[0].player4 = {
      bet: 0,
      got: 0,
    };

    const element = renderer.create(
            <RunningScores game={game} />,
    ).toJSON();

    expect(element).toMatchSnapshot();
  });

  it('renders correctly with 2 players and multiple rounds', () => {
    const game = { ...baseGame };
    game.rounds = [...game.rounds, {
      roundNumber: 2,
      player1: {
        bet: 20,
        got: 20,
      },
      player2: {
        bet: 15,
        got: 30,
      },
    }];

    const element = renderer.create(
            <RunningScores game={game} />,
    ).toJSON();

    expect(element).toMatchSnapshot();
  });

  it('renders correctly with 3 players and multiple rounds', () => {
    const game = { ...baseGame };
    game.players.player3 = 'Player3';
    game.rounds[0].player3 = {
      bet: 20,
      got: 20,
    };
    game.rounds = [...game.rounds, {
      roundNumber: 2,
      player1: {
        bet: 20,
        got: 20,
      },
      player2: {
        bet: 15,
        got: 30,
      },
      player3: {
        bet: 0,
        got: 0,
      },
    }];

    const element = renderer.create(
            <RunningScores game={game} />,
    ).toJSON();

    expect(element).toMatchSnapshot();
  });

  it('renders correctly with 4 players and multiple rounds', () => {
    const game = { ...baseGame };
    game.players.player3 = 'Player3';
    game.players.player4 = 'Player4';
    game.rounds[0].player3 = {
      bet: 20,
      got: 20,
    };
    game.rounds[0].player4 = {
      bet: 0,
      got: 0,
    };
    game.rounds = [...game.rounds, {
      roundNumber: 2,
      player1: {
        bet: 20,
        got: 20,
      },
      player2: {
        bet: 15,
        got: 30,
      },
      player3: {
        bet: 0,
        got: 0,
      },
      player4: {
        bet: 18,
        got: 17,
      },
    }];

    const element = renderer.create(
            <RunningScores game={game} />,
    ).toJSON();

    expect(element).toMatchSnapshot();
  });
});
