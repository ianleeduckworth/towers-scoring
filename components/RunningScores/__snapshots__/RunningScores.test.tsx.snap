// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button renders correctly with 2 players and 1 round 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F0F0F0",
      "borderBottomColor": "#000000",
      "borderTopColor": "#000000",
      "borderWidth": 1,
      "height": 100,
      "paddingLeft": 40,
      "paddingRight": 40,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 3,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 4,
        }
      }
    >
      Scores
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player1
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        3
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player2
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        3
      </Text>
    </View>
  </View>
</View>
`;

exports[`button renders correctly with 2 players and multiple rounds 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F0F0F0",
      "borderBottomColor": "#000000",
      "borderTopColor": "#000000",
      "borderWidth": 1,
      "height": 100,
      "paddingLeft": 40,
      "paddingRight": 40,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 3,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 4,
        }
      }
    >
      Scores
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player1
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        35
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player2
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        6
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player3
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        32
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player4
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        50
      </Text>
    </View>
  </View>
</View>
`;

exports[`button renders correctly with 3 players and 1 round 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F0F0F0",
      "borderBottomColor": "#000000",
      "borderTopColor": "#000000",
      "borderWidth": 1,
      "height": 100,
      "paddingLeft": 40,
      "paddingRight": 40,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 3,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 4,
        }
      }
    >
      Scores
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player1
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        3
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player2
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        3
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player3
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        32
      </Text>
    </View>
  </View>
</View>
`;

exports[`button renders correctly with 3 players and multiple rounds 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F0F0F0",
      "borderBottomColor": "#000000",
      "borderTopColor": "#000000",
      "borderWidth": 1,
      "height": 100,
      "paddingLeft": 40,
      "paddingRight": 40,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 3,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 4,
        }
      }
    >
      Scores
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player1
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        35
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player2
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        6
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player3
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        82
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player4
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        50
      </Text>
    </View>
  </View>
</View>
`;

exports[`button renders correctly with 4 players and 1 round 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F0F0F0",
      "borderBottomColor": "#000000",
      "borderTopColor": "#000000",
      "borderWidth": 1,
      "height": 100,
      "paddingLeft": 40,
      "paddingRight": 40,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 3,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 4,
        }
      }
    >
      Scores
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player1
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        3
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player2
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        3
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player3
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        32
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player4
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        50
      </Text>
    </View>
  </View>
</View>
`;

exports[`button renders correctly with 4 players and multiple rounds 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F0F0F0",
      "borderBottomColor": "#000000",
      "borderTopColor": "#000000",
      "borderWidth": 1,
      "height": 100,
      "paddingLeft": 40,
      "paddingRight": 40,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 3,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 4,
        }
      }
    >
      Scores
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player1
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        35
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player2
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        6
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player3
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        82
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "column",
          "paddingRight": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player4
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        68
      </Text>
    </View>
  </View>
</View>
`;
